From ec6dfcc322e94a6ce20c99f664a062644aaea469 Mon Sep 17 00:00:00 2001
From: Haihao Xiang <haihao.xiang@intel.com>
Date: Wed, 21 Aug 2019 12:45:37 +0800
Subject: [PATCH 1/3] msdkvpp: free all msdk frames before closing VPP session

Otherwise the new VPP session will still use the old external frame
allocator and might result in some memory leaks
---
 sys/msdk/gstmsdkvpp.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/sys/msdk/gstmsdkvpp.c b/sys/msdk/gstmsdkvpp.c
index d0e0a46..4fbe855 100644
--- a/sys/msdk/gstmsdkvpp.c
+++ b/sys/msdk/gstmsdkvpp.c
@@ -781,6 +781,11 @@ gst_msdkvpp_close (GstMsdkVPP * thiz)
         msdk_status_to_string (status));
   }
 
+  if (thiz->use_video_memory) {
+    gst_msdk_frame_free (thiz->context, &thiz->in_alloc_resp);
+    gst_msdk_frame_free (thiz->context, &thiz->out_alloc_resp);
+  }
+
   if (thiz->context)
     gst_object_replace ((GstObject **) & thiz->context, NULL);
 
@@ -927,9 +932,15 @@ gst_msdkvpp_initialize (GstMsdkVPP * thiz)
    * otherwise the subsequent function call of MFXVideoVPP_Init() will
    * fail
    */
-  if (thiz->initialized)
+  if (thiz->initialized) {
     MFXVideoVPP_Close (session);
 
+    if (thiz->use_video_memory) {
+      gst_msdk_frame_free (thiz->context, &thiz->in_alloc_resp);
+      gst_msdk_frame_free (thiz->context, &thiz->out_alloc_resp);
+    }
+  }
+
   if (thiz->use_video_memory) {
     gst_msdk_set_frame_allocator (thiz->context);
     thiz->param.IOPattern =
-- 
2.22.0

